src/com/ms/silverking/cloud/meta/ChildrenWatcher.java

```diff
@@ -17,101 +17,102 @@
+      long intervalMillis, long maxInitialSleep) {
+    super(metaClientCore, timer, basePath, intervalMillis, maxInitialSleep);
+    this.listener = listener;
+  }
+
+  public ChildrenWatcher(MetaClientCore metaClientCore, String basePath, ChildrenListener listener, long intervalMillis,
+      long maxInitialSleep) {
+    this(null, metaClientCore, basePath, listener, intervalMillis, maxInitialSleep);
+  }
+
+  public ChildrenWatcher(Timer timer, MetaClientCore metaClientCore, String basePath, ChildrenListener listener,
+      long intervalMillis) {
+    this(timer, metaClientCore, basePath, listener, intervalMillis, intervalMillis);
+  }
+
+  public ChildrenWatcher(MetaClientCore metaClientCore, String basePath, ChildrenListener listener,
+      long intervalMillis) {
+    this(metaClientCore, basePath, listener, intervalMillis, intervalMillis);
+  }
+
+  private Map<String, byte[]> readChildStates() throws KeeperException {
+    ZooKeeperExtended _zk;
+    List<String> children;
+
+    _zk = metaClientCore.getZooKeeper();
+    children = _zk.getChildren(basePath, this);
+    return _zk.getByteArrays(basePath, ImmutableSet.copyOf(children), this, this);
+  }
+
+  protected void _doCheck() throws KeeperException {
+    Map<String, byte[]> latestChildStates;
+
+    latestChildStates = readChildStates();
+    if (!mapsAreEqual(latestChildStates, childStates)) {
+      childStates = latestChildStates;
+      listener.childrenChanged(basePath, latestChildStates);
     }
+  }
 
-    private void checkChildStates() {
-        try {
-            doCheck();
-        } catch (KeeperException ke) {
-            throw new RuntimeException(ke);
-        }
+  private void checkChildStates() {
+    try {
+      doCheck();
+    } catch (KeeperException ke) {
+      throw new RuntimeException(ke);
     }
-    
-    /**
-     * Can't use Guava Maps.difference() since we have byte arrays here.
-     * @param latestChildStates
-     * @param childStates2
-     * @return
-     */
-    private boolean mapsAreEqual(Map<String, byte[]> m1, Map<String, byte[]> m2) {
-        Set<String> m1KeySet;
-        
-        m1KeySet = m1.keySet();
-        if (m2 != null && m1KeySet.equals(m2.keySet())) {
-            for (String key : m1KeySet) {
-                if (!Arrays.equals(m1.get(key), m2.get(key))) {
-                    return false;
-                }
-            }
-            return true;
-        } else {
-            return false;
+  }
+
+  /**
+   * Can't use Guava Maps.difference() since we have byte arrays here.
+   *
+   * @param latestChildStates
+   * @param childStates2
+   * @return
+   */
+  private boolean mapsAreEqual(Map<String, byte[]> m1, Map<String, byte[]> m2) {
+    Set<String> m1KeySet;
+
+    m1KeySet = m1.keySet();
+    if (m2 != null && m1KeySet.equals(m2.keySet())) {
+      for (String key : m1KeySet) {
+        if (!Arrays.equals(m1.get(key), m2.get(key))) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.